picatrix/lib/framework.py

```diff
@@ -0,0 +1,494 @@
+      arg_type = str
+    elif arg_type_string == 'float':
+      arg_type = float
+    elif arg_type_string == 'int':
+      arg_type = int
+    elif arg_type_string == 'bool':
+      arg_type = bool
+      arg_default = default_values.get(variable)
+      if arg_default:
+        action = 'store_false'
+      else:
+        action = 'store_true'
+    else:
+      # Type is a string reference to an object.
+      arg_type = str
+      parser.storage[variable] = arg_type_string
+
+    if arg_type == bool:
+      parser.add_argument(
+          '--{0:s}'.format(variable), '-{0:s}'.format(variable),
+          dest=variable, action=action,
+          default=default_values.get(variable),
+          help=description)
+    else:
+      parser.add_argument(
+          '--{0:s}'.format(variable), '-{0:s}'.format(variable),
+          dest=variable, type=arg_type, action='store',
+          default=default_values.get(variable),
+          help=description)
+
+
+def _parse_docstring(function: MagicProtocol) -> List[Dict[str, str]]:
+  """Return a list of arguments extracted from a function's docstring.
+
+  Args:
+    function (function): a function to extract the docstring from.
+
+  Returns:
+    A list of dict, one per argument. Each dict will have the following
+    attributes; variable, type and description.
+  """
+  doc_string = getattr(function, '__doc__')
+  if not doc_string:
+    return []
+
+  doc_lines = doc_string.split('\n')
+  arg_lines = _get_argument_lines_from_docstring(doc_lines)
+  return _get_arguments_from_arg_lines(arg_lines)
+
+
+def _parse_line_string(line: str) -> List[str]:
+  """Return a list of arguments from a line magic.
+
+  Args:
+    line (str): the value passed to the line magic, or line attribute in a
+        cell magic.
+
+  Returns:
+    List of arguments that can be parsed by argparse.
+  """
+  arguments = []
+  temp_args = []
+  in_quotes = False
+  arg_item = False
+  quota_char = ''
+
+  for item in line.strip().split():
+    # Two cases to watch for, line starts with - (argument) or quotes.
+    if in_quotes:
+      if item[-1] == quota_char:
+        temp_args.append(item[:-1])
+        arguments.append(' '.join(temp_args))
+        temp_args = []
+        in_quotes = False
+        quota_char = ''
+      else:
+        temp_args.append(item)
+      continue
+
+    if item[0] == '-':
+      arguments.append(item)
+      arg_item = True
+      continue
+
+    if item[0] in ('\'', '"'):
+      quota_char = item[0]
+
+      # The quoted argument is a single word.
+      if item[-1] == quota_char:
+        arguments.append(item[1:-1])
+        quota_char = ''
+        continue
+
+      temp_args.append(item[1:])
+      in_quotes = True
+      quota_char = item[0]
+      continue
+
+    if arg_item:
+      arguments.append(item)
+      arg_item = False
+    else:
+      temp_args.append(item)
+
+  if temp_args:
+    arguments.append(' '.join(temp_args))
+
+  return arguments
+
+
+def picatrix_magic(
```

Comment:
to avoid confusion with other functions, this si what is used to register a function as a magic, so you have


```
@framework.picatrix_magic
def my_magical_magic(data, foo, bar):
...
```

I think being explicit about the name here is fine.

Reply:
Ack